twoliter/src/lock/mod.rs

```diff
@@ -0,0 +1,262 @@
+        let lock_str = toml::to_string(&lock_state).context("failed to serialize lock file")?;
+
+        debug!("Writing new lock file to '{}'", lock_file_path.display());
+        write(&lock_file_path, lock_str)
+            .await
+            .context("failed to write lock file")?;
+        Ok(lock_state)
+    }
+
+    #[instrument(level = "trace", skip(project))]
+    pub(crate) async fn load(project: &Project) -> Result<Self> {
+        let lock_file_path = project.project_dir().join(TWOLITER_LOCK);
+        ensure!(
+            lock_file_path.exists(),
+            "Twoliter.lock does not exist, please run `twoliter update` first"
+        );
+        debug!("Loading existing lockfile '{}'", lock_file_path.display());
+        let lock_str = read_to_string(&lock_file_path)
+            .await
+            .context("failed to read lockfile")?;
+        let lock: Self =
+            toml::from_str(lock_str.as_str()).context("failed to deserialize lockfile")?;
+        info!("Resolving project references to check against lock file");
+        let lock_state = Self::resolve(project).await?;
+
+        ensure!(lock_state == lock, "changes have occured to Twoliter.toml or the remote kit images that require an update to Twoliter.lock");
+        Ok(lock)
+    }
+
+    fn external_kit_metadata(&self) -> ExternalKitMetadata {
+        ExternalKitMetadata {
+            sdk: self.sdk.clone(),
+            kits: self.kit.clone(),
+        }
+    }
+
+    /// Fetches all external kits defined in a Twoliter.lock to the build directory
+    #[instrument(level = "trace", skip_all)]
+    pub(crate) async fn fetch(&self, project: &Project, arch: &str) -> Result<()> {
+        let image_tool = ImageTool::from_environment()?;
+        let target_dir = project.external_kits_dir();
+        create_dir_all(&target_dir).await.context(format!(
+            "failed to create external-kits directory at {}",
+            target_dir.display()
+        ))?;
+
+        info!(
+            dependencies = ?self.kit.iter().map(ToString::to_string).collect::<Vec<_>>(),
+            "Extracting kit dependencies."
+        );
+        for image in self.kit.iter() {
+            let vendor = project
+                .vendor()
+                .get(&ValidIdentifier(image.vendor.clone()))
+                .context(format!(
+                    "failed to find vendor for kit with name '{}' and vendor '{}'",
+                    image.name, image.vendor
+                ))?;
+            let override_ = project
+                .overrides()
+                .get(&image.vendor)
+                .and_then(|x| x.get(&image.name));
+            let resolver =
+                ImageResolver::from_locked_image(image, image.vendor.as_str(), vendor, override_);
+            resolver
+                .extract(&image_tool, &project.external_kits_dir(), arch)
+                .await?;
+        }
+        let mut kit_list = Vec::new();
```

Comment:
I'd still prefer if we refactor this to minimize the overall complexity of this function.